homeassistant/components/ads/sensor.py

```diff
@@ -84,6 +123,15 @@ async def async_added_to_hass(self) -> None:
         )
 
     @property
-    def native_value(self) -> StateType:
+    def native_value(self) -> StateType | datetime | date:
         """Return the state of the device."""
-        return self._state_dict[STATE_KEY_STATE]
+        state = self._state_dict.get(STATE_KEY_STATE)
+
+        if self._device_class == SensorDeviceClass.TIMESTAMP and isinstance(state, int):
+            # Convert integer timestamp to datetime object
+            return datetime.fromtimestamp(state, tz=UTC)
+        if self._device_class == SensorDeviceClass.DATE and isinstance(state, int):
+            # Convert integer timestamp to datetime object for date only
+            dt = datetime.fromtimestamp(state, tz=UTC)
+            return dt.date()
```

Comment:
Does it make sense to add the conversion in the place where we fetch the data instead of here? Because we know if the datatype is a date right?

Reply:
Oh that isn't the best place as its now a polling sensor. I mean, the previous PR you used logic to make string a string, bool a bool, etc, can't we move this to there?

Reply:
tried it, but can't really get my head around it.... removing it for now (not the most useful types anyways. will add if i have more time.)

Reply:
I think there are two ways to look at this...

Option 1 (as previously implemented):
- in `def _device_notification_callback`, keep the raw integer value
- in `def native_value`, ignore the `ADSTYPE` and use the `SensorDeviceClass` to convert the value to correct datetime

Option 2 (I think this is what @joostlek has in mind)
- in `def _device_notification_callback`, use `ADSTYPE` to convert to date/time/datetime/timestamp
- in `def native_value`, return the existing datetime value from the state_dict
- probably need to add some checks in `setup_platform` to confirm that `DeviceClass` is compatible with `ADSTYPE`

Reply:
Will try to achieve option 2 in a future PR. Now, i have a few more ADS devices to add, and would like to focus on those.